File: wiki-backend/src/main/java/com/internal/team/wiki/user/domain/UserEntity.java
package com.internal.team.wiki.user.domain;

import com.internal.team.wiki.global.BaseEntityTime;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import java.util.Objects;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Table(name = "users")
@Entity
public class UserEntity extends BaseEntityTime {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Embedded
    private Username username;

    @Embedded
    private Nickname nickname;

    @Embedded
    private Password password;

    @Enumerated(EnumType.STRING)
    private RoleType roleType = RoleType.USER;

    @Builder
    public UserEntity(final Username username, final Nickname nickname, final Password password) {
        this.username = username;
        this.nickname = nickname;
        this.password = password;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof UserEntity)) return false;
        UserEntity that = (UserEntity) o;
        return Objects.equals(getId(), that.getId());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getId());
    }
}

File: wiki-backend/src/main/java/com/internal/team/wiki/user/domain/UserEntity.kt
package com.internal.team.wiki.user.domain

import com.internal.team.wiki.global.BaseTimeEntity
import java.time.LocalDateTime
import javax.persistence.*

@Table(name = "users")
@Entity
class UserEntity(

    @Embedded
    var username: Username,

    @Embedded
    val nickname: Nickname,

    @Embedded
    var password: Password,

    @Enumerated(EnumType.STRING)
    val roleType: RoleType = RoleType.USER,

    @Column(name = "deleted", nullable = false)
    var deleted: Boolean = false,

    @Column(name = "scheduled_deletion_time", nullable = true)
    var scheduledDeletionTime: LocalDateTime?,

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null
) : BaseTimeEntity() {

    fun updateDeleted() {
        this.deleted = true
    }

    fun scheduleDeletion(scheduledDeletionTime: LocalDateTime) {
        this.scheduledDeletionTime = scheduledDeletionTime
    }
}